ui/src/state/mod.rs

```diff
@@ -501,6 +502,68 @@ impl State {
     pub fn remove_window(&mut self, id: WindowId) {
         self.ui.remove_overlay(id);
     }
+
+    pub fn add_message_reaction(&mut self, chat_id: Uuid, message_id: Uuid, emoji: String) {
+        let user = self.account.identity.did_key();
+        let conv = match self.chats.all.get_mut(&chat_id) {
+            Some(c) => c,
+            None => {
+                log::warn!("attempted to add reaction to nonexistent conversation");
+                return;
+            }
+        };
+
+        for msg in &mut conv.messages {
```

Comment:
As I look at this section, You could probably get away with doing something like 

```rust
if let Some(msg) = conv
            .messages
            .iter()
            .position(|msg| msg.id() == message_id)
            .and_then(|index| conv.messages.get_mut(index))
        {
            let mut has_emoji = false;
            for reaction in msg.reactions_mut() {
                if !reaction.emoji().eq(&emoji) {
                    continue;
                }
                if !reaction.users().contains(&user) {
                    reaction.users_mut().push(user.clone());
                    has_emoji = true;
                }
            }

            if !has_emoji {
                let mut r = Reaction::default();
                r.set_emoji(&emoji);
                r.set_users(vec![user]);
                msg.reactions_mut().push(r);
            }
        }
```

Or instead you could omit `and_then` and switch the iterator to mutable and use `find` which should return a mutable message, if it is found. Eg

```rust
if let Some(msg) = conv
            .messages
            .iter_mut()
            .find(|msg| msg.id() == message_id)
        {
            let mut has_emoji = false;
            for reaction in msg.reactions_mut() {
                if !reaction.emoji().eq(&emoji) {
                    continue;
                }
                if !reaction.users().contains(&user) {
                    reaction.users_mut().push(user.clone());
                    has_emoji = true;
                }
            }

            if !has_emoji {
                let mut r = Reaction::default();
                r.set_emoji(&emoji);
                r.set_users(vec![user]);
                msg.reactions_mut().push(r);
            }
        }
```